sw-emulator/lib/periph/src/ml_dsa87.rs

```diff
@@ -0,0 +1,1254 @@
+        _val: RvData,
+    ) -> Result<(), BusError> {
+        Err(BusError::StoreAccessFault)
+    }
+
+    // Clear registers
+    fn zeroize(&mut self) {}
+
+    /// On Write callback for `control` register
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - Size of the write
+    /// * `val` - Data to write
+    ///
+    /// # Error
+    ///
+    /// * `BusError` - Exception with cause `BusError::StoreAccessFault` or `BusError::StoreAddrMisaligned`
+    pub fn on_write_control(&mut self, size: RvSize, val: RvData) -> Result<(), BusError> {
+        // Writes have to be Word aligned
+        if size != RvSize::Word {
+            Err(BusError::StoreAccessFault)?
+        }
+
+        // Set the control register
+        self.control.reg.set(val);
+
+        match self.control.reg.read_as_enum(Control::CTRL) {
+            Some(Control::CTRL::Value::GEN_KEY)
+            | Some(Control::CTRL::Value::SIGN)
+            | Some(Control::CTRL::Value::VERIFY) => {
+                // Reset the Ready and Valid status bits
+                self.status
+                    .reg
+                    .modify(Status::READY::CLEAR + Status::VALID::CLEAR);
+
+                self.op_complete_action = Some(self.timer.schedule_poll_in(ML_DSA87_OP_TICKS));
+            }
+            _ => {}
+        }
+
+        if self.control.reg.is_set(Control::ZEROIZE) {
+            self.zeroize();
+        }
+
+        Ok(())
+    }
+
+    fn gen_key(&mut self) {
+        let seed_bytes = &bytes_from_words_le(&self.seed);
+        let mut rng = StdRng::from_seed(*seed_bytes);
+        let (pk, sk) = try_keygen_with_rng(&mut rng).unwrap();
+
+        self.pk = words_from_bytes_le(&pk.into_bytes());
+        self.sk_out = words_from_bytes_le(&sk.into_bytes());
+    }
+
+    fn sign(&mut self) {
+        let seed_bytes = &bytes_from_words_le(&self.seed);
+        let mut rng = StdRng::from_seed(*seed_bytes);
+
+        let secret_key_bytes = &bytes_from_words_le(&self.sk_in);
+        let secret_key = PrivateKey::try_from_bytes(*secret_key_bytes).unwrap();
+
+        let message = &bytes_from_words_le(&self.message);
+
+        // The Ml_Dsa87 5 signature is 4595 len but the reg is one byte longer
+        let signature = secret_key.try_sign_with_rng(&mut rng, message).unwrap();
+        let signature_extended = {
+            let mut sig = [0; SIG_LEN + 1];
+            sig[..SIG_LEN].copy_from_slice(&signature);
+            sig
+        };
+        self.signature
+            .copy_from_slice(&words_from_bytes_le(&signature_extended));
+    }
+
+    fn verify(&mut self) {
+        let message = &bytes_from_words_le(&self.message);
+
+        let public_key = {
+            let key_bytes = &bytes_from_words_le(&self.pk);
+            PublicKey::try_from_bytes(*key_bytes).unwrap()
+        };
+
+        let signature = &bytes_from_words_le(&self.signature);
+
+        let result = public_key.verify(message, &signature[..SIG_LEN].try_into().unwrap());
```

Comment:
Shouldn't signature be a fixed size? Not sure why we need to take a subslice

Reply:
I think this is because `self.signature` is `[u32; 1157]` which expands to `[u8; 4628]`, but the `SIG_LEN` is 4627.

Reply:
Ah, makes sense. Thanks :)